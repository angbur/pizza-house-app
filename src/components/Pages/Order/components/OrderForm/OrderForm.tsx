import Button from 'components/Atoms/Button/Button';
import InputWithLabel from 'components/Atoms/InputWithLabel/InputWithLabel';
import Typography from 'components/Atoms/Typography/Typography';
import { useNavigate } from 'react-router-dom';
import styled from 'styled-components';

const Container = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  text-align: center;
  width: 100%;
  & > h2 {
    margin: 2rem 0;
  }
  margin: 2rem;
`;

const Box  = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1.2rem;
`;

const FormActions = styled.div`
  display: flex;
  gap: 3rem;
  margin: 2rem;
`;

const OrderForm = () => {
  const navigate = useNavigate();

  const handleBack = () => {
    navigate('/order');
  };

  const handleNext = () => {
    navigate('/order/summary');
  };

  return (
    <Container>
      <Typography variant='section-title' color='light'>
        Form
      </Typography>
      <form>
        <Box>
          <InputWithLabel label={'Name'} type={'text'} placeholder={'Your name'}/>
          <InputWithLabel label={'Email'} type={'email'} placeholder={'Your email'}/>
          <InputWithLabel label={'City'} type={'text'} placeholder={'Your city'}/>
          <InputWithLabel label={'Postal Code'} type={'text'} placeholder={'00-000'}/>
          <InputWithLabel label={'Street'} type={'text'} placeholder={'Street'}/>
          <InputWithLabel label={'Apartment/house number'} type={'text'} placeholder={'Number'}/>
          <InputWithLabel label={'Phone'} type={'tel'} placeholder={'000-000-000'}/>
        </Box>
      </form>
      <FormActions>
          <Button variant='button-text-light' onClick={handleBack}>
            Back
          </Button>
          <Button variant='primary-light' onClick={handleNext}>Summary</Button>
      </FormActions>
    </Container>
  );
};

export default OrderForm;
