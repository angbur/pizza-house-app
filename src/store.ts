import { configureStore } from '@reduxjs/toolkit';
import orderReducer from 'components/Pages/Order/orderSlice';
import { pizzaApi } from 'services/pizza';

export const store = configureStore({
  reducer: {
    order: orderReducer,
    [pizzaApi.reducerPath]: pizzaApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(pizzaApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
