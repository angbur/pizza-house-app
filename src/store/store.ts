import { configureStore, Store } from '@reduxjs/toolkit';
import orderReducer from 'components/Pages/Order/orderSlice';
import dialogReducer from 'components/Organisms/Dialog/dialogSlice';
import { pizzaApi } from 'services/pizza';
import orderMiddleware from './middleware/orderMiddleware';

export const store: Store = configureStore({
  reducer: {
    order: orderReducer,
    dialog: dialogReducer,
    [pizzaApi.reducerPath]: pizzaApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(pizzaApi.middleware, orderMiddleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
