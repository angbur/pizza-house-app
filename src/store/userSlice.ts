import { createSlice } from '@reduxjs/toolkit';
import { RootState } from './store';
import { toast } from 'react-toastify';
import { userApi } from 'services/user';

type User = {
  loggedIn: boolean;
  status: string | undefined;
  error: string | undefined;
  userId: string | undefined;
  login: string | undefined;
  token: string | undefined;
};

const initialState: User = {
  loggedIn: false,
  status: 'idle',
  error: undefined,
  userId: undefined,
  login: undefined,
  token: undefined,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logout: (state) => {
      state.loggedIn = false;
      state.userId = undefined;
      state.login = undefined;
      state.token = undefined;
      toast.success('Logged out!');
    },
  },
  extraReducers: (builder) => {
    builder.addMatcher(userApi.endpoints.login.matchFulfilled, (state, { payload }) => {
      state.loggedIn = true;
      state.token = payload.token;
      state.userId = payload.userId;
      state.login = payload.login;
    });
  },
});

export const { logout } = userSlice.actions;

export const selectName = (state: RootState) => state.user.user;

export const isLoggedIn = (state: RootState) => state.user.loggedIn;

export const selectResponseStatus = (state: RootState) => state.user.status;

export const selectResponseError = (state: RootState) => state.user.error;

export const selectLoggedInUser = (state: RootState) => state.user.user;

export default userSlice.reducer;
