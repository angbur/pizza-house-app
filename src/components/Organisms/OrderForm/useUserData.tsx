import { useAppSelector } from 'hooks';
import { useState, useEffect } from 'react';
import { useGetUserByIdQuery } from 'services/user/user';
import { selectUserId } from 'store/user/userSlice';

export type ImportedUserData = {
  firstName: string;
  lastName: string;
  email: string;
};

export const useUserData = (): ImportedUserData => {
  const userId: string | undefined = useAppSelector(selectUserId);
  const user = useGetUserByIdQuery(userId as string);
  const [formData, setData] = useState<ImportedUserData>({
    firstName: '',
    lastName: '',
    email: '',
  });

  useEffect(() => {
    setData({
      firstName: user.data?.firstName ?? '',
      lastName: user.data?.lastName ?? '',
      email: user.data?.email ?? '',
    });
  }, [user]);

  return formData;
};
