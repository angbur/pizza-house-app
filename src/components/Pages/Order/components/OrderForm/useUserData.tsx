import { useAppSelector } from 'hooks';
import { useState, useEffect } from 'react';
import { useGetUserByIdQuery } from 'services/user';
import { selectUserId } from 'store/userSlice';

export type ImportedUserData = {
    firstName: string;
    lastName: string;
    email: string;
  };

export const useUserData = (): ImportedUserData => {
    const userId = useAppSelector(selectUserId);
    const user = useGetUserByIdQuery(userId);
    const [formData, setData] = useState<ImportedUserData>({
        firstName: '',
        lastName: '',
        email: '',
      });

      useEffect(() => {
        setData({
            firstName: user.data?.firstName ?? '',
            lastName: user.data?.lastName ?? '',
            email: user.data?.email ?? '',
        })
      }, [user]);

      return formData;
};