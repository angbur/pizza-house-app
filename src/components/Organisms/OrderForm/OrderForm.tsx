import Button from 'components/Atoms/Button/Button';
import InputWithLabel from 'components/Atoms/InputWithLabel/InputWithLabel';
import Typography from 'components/Atoms/Typography/Typography';
import { useAppDispatch, useAppSelector } from 'hooks';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import { useCreateOrderMutation } from 'services/order';
import { selectUserId } from 'store/userSlice';
import styled from 'styled-components';
import { OrderItem } from 'types/Order';
import {
  removeAllOrder,
  selectListOfOrderItems,
  selectSumOfOrder,
} from '../../Pages/Order/orderSlice';
import OrderList from '../OrderList/OrderList';
import OrderSummary from '../OrderSummary/OrderSummary';
import { ImportedUserData } from './useUserData';

const Container = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  text-align: center;
  width: 100%;
  & > h2 {
    margin: 2rem 0 4rem 0;
  }
  margin: 2rem;
  margin-bottom: 10rem;
`;

const Box = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1.2rem;
`;

const FormActions = styled.div`
  display: flex;
  gap: 3rem;
  margin: 2rem;
  @media (max-width: 560px) {
    flex-direction: column-reverse;
  }
`;

const SummaryContainer = styled.div`
  display: flex;
  justify-content: space-around;
  width: 100%;
  padding: 0 2rem;
  gap: 1rem;
  @media (max-width: 766px) {
    flex-direction: column;
    justify-content: center;
    align-items: center;
    padding: 0;
  }
`;

const StyledDiv = styled.div`
  display: flex;
  gap: 3rem;
  margin: 4rem;
  @media (max-width: 560px) {
    flex-direction: column-reverse;
    margin-bottom: 10rem;
  }
`;

type OrderForm = {
  firstName: string;
  lastName: string;
  email: string;
  city: string;
  postalCode: string;
  street: string;
  houseNumber: string;
  phone: string;
};

type OrderFormProps = {
  user: ImportedUserData;
};

const OrderForm = ({ user }: OrderFormProps) => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const userId: string = useAppSelector(selectUserId);
  const orderList: OrderItem[] = useAppSelector(selectListOfOrderItems);
  const sum: number = useAppSelector(selectSumOfOrder);
  const [isValidable, setValidation] = useState<boolean>(false);
  const [formData, setData] = useState<OrderForm>({
    firstName: user.firstName,
    lastName: user.lastName,
    email: user.email,
    city: '',
    postalCode: '',
    street: '',
    houseNumber: '',
    phone: '',
  });
  const [createOrder] = useCreateOrderMutation();

  const handleBack = () => {
    navigate('/order');
  };

  const handleSubmit = () => {
    for (const value in formData) {
      if (formData[value as keyof typeof formData].length < 1) {
        toast.error('Please fill in the missing fields');
        setValidation(false);
        break;
      } else {
        setValidation(true);
      }
    }
  };

  const handleChange = ({ target: { name, value } }: React.ChangeEvent<HTMLInputElement>) => {
    setData({
      ...formData,
      [name.trim()]: value,
    });
  };

  const handleCancel = () => {
    dispatch(removeAllOrder());
    navigate('/');
  };

  const handleEdit = () => {
    setValidation((prev) => !prev);
  };

  const handleOrder = async () => {
    const order: Omit<OrderItem, '_id'>[] = [];

    for (const element of orderList) {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { _id, ...rest } = element;
      order.push({ ...rest });
    }

    const requestData = {
      user: userId,
      order: order,
      deliveryAddress: formData,
      total: sum,
    };

    try {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const order = await createOrder(requestData).unwrap();
      dispatch(removeAllOrder());
      toast.success('Order sent!');
      navigate('/');
    } catch (err) {
      console.log(err);
    }
  };

  return (
    <Container>
      <Typography variant='section-title' color='light'>
        {isValidable ? 'Summary' : 'Form'}
      </Typography>
      {!isValidable ? (
        <>
          <form>
            <Box>
              <InputWithLabel
                label={'Name'}
                name={'firstName'}
                type={'text'}
                placeholder={'Your first name'}
                color={'light'}
                value={formData.firstName}
                onChange={handleChange}
                error={!/^[A-Za-zżźćńółęąśŻŹĆĄŚĘŁÓŃ]{2,50}$/.test(formData.firstName.trim())}
              />
              <InputWithLabel
                label={'Last Name'}
                name={'lastName'}
                type={'text'}
                placeholder={'Your last name'}
                color={'light'}
                value={formData.lastName}
                onChange={handleChange}
                error={!/^[A-Za-zżźćńółęąśŻŹĆĄŚĘŁÓŃ]{2,50}$/.test(formData.lastName.trim())}
              />
              <InputWithLabel
                label={'Email'}
                name={'email'}
                type={'email'}
                placeholder={'Your email'}
                color={'light'}
                value={formData.email}
                onChange={handleChange}
                error={
                  !/^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/.test(
                    formData.email.trim(),
                  )
                }
              />
              <InputWithLabel
                label={'City'}
                name={'city'}
                type={'text'}
                placeholder={'Your city'}
                color={'light'}
                value={formData.city}
                onChange={handleChange}
                error={
                  !/^([a-zA-Z\u0080-\u024F]+(?:. |-| |'))*[a-zA-Z\u0080-\u024F]{3,20}$/.test(
                    formData.city.trim(),
                  )
                }
              />
              <InputWithLabel
                label={'Postal Code'}
                name={'postalCode'}
                type={'text'}
                placeholder={'00-000'}
                color={'light'}
                value={formData.postalCode}
                onChange={handleChange}
                error={!/^\d{2}((-?)\d{3})$/.test(formData.postalCode.trim())}
              />
              <InputWithLabel
                label={'Street'}
                name={'street'}
                type={'text'}
                placeholder={'Street'}
                color={'light'}
                value={formData.street}
                onChange={handleChange}
                error={!/^[A-Za-zżźćńółęąśŻŹĆĄŚĘŁÓŃ]{2,40}$/.test(formData.street.trim())}
              />
              <InputWithLabel
                label={'House / apartment number '}
                name={'houseNumber'}
                type={'text'}
                placeholder={'Number'}
                color={'light'}
                value={formData.houseNumber}
                onChange={handleChange}
                error={formData.houseNumber.trim().length === 0}
              />
              <InputWithLabel
                label={'Phone'}
                name={'phone'}
                type={'tel'}
                placeholder={'000-000-000'}
                color={'light'}
                value={formData.phone}
                onChange={handleChange}
                error={!/^\d{3}(-?)\d{3}(-?)\d{3}$/.test(formData.phone.trim())}
              />
            </Box>
          </form>
          <FormActions>
            <Button variant='button-text-light' onClick={handleBack}>
              Back
            </Button>
            <Button variant='primary-light' onClick={handleSubmit}>
              Summary
            </Button>
          </FormActions>
        </>
      ) : (
        <>
          <SummaryContainer>
            <OrderSummary data={formData} onClick={handleEdit} />
            <OrderList />
          </SummaryContainer>
          <StyledDiv>
            <Button variant='button-text-light' onClick={handleCancel}>
              Cancel
            </Button>
            <Button variant='primary-light' onClick={handleOrder}>
              Order now
            </Button>
          </StyledDiv>
        </>
      )}
    </Container>
  );
};

export default OrderForm;
