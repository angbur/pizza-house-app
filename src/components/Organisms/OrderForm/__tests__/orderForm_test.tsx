import renderer from 'react-test-renderer';
import ThemeProvider from 'components/Theme/ThemeContext';
import { store } from 'store/store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import OrderForm from '../OrderForm';

const Wrapper = () => {
  return (
    <ThemeProvider>
      <BrowserRouter>
        <Provider store={store}>
          <OrderForm
            user={{
              firstName: 'Kamila',
              lastName: 'Lasocka',
              email: 'lasocka@gmail.com',
            }}
          />
        </Provider>
      </BrowserRouter>
    </ThemeProvider>
  );
};

describe('Order Form', () => {
  it('should renders order form correctly', () => {
    const tree = renderer.create(<Wrapper />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
