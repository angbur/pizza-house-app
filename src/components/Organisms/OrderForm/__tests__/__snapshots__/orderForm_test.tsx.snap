// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Form should renders order form correctly 1`] = `
<section
  className="sc-dIfARi bImcQW"
>
  <h2
    className="section-title"
    style={
      Object {
        "color": "#FCFBFB",
        "fontFamily": "Source Sans Pro",
        "fontSize": "2.25rem",
        "fontWeight": "300",
        "letterSpacing": "1px",
        "lineHeight": "1.3",
      }
    }
  >
    Form
  </h2>
  <form>
    <div
      className="sc-hHTYSt ftghfI"
    >
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#FCFBFB",
            }
          }
        >
          Name
        </label>
        <input
          name="firstName"
          onChange={[Function]}
          placeholder="Your first name"
          style={
            Object {
              "color": "#151515",
            }
          }
          type="text"
          value="Kamila"
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#FCFBFB",
            }
          }
        >
          Last Name
        </label>
        <input
          name="lastName"
          onChange={[Function]}
          placeholder="Your last name"
          style={
            Object {
              "color": "#151515",
            }
          }
          type="text"
          value="Lasocka"
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#FCFBFB",
            }
          }
        >
          Email
        </label>
        <input
          name="email"
          onChange={[Function]}
          placeholder="Your email"
          style={
            Object {
              "color": "#151515",
            }
          }
          type="email"
          value="lasocka@gmail.com"
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#e74c3c",
            }
          }
        >
          City
        </label>
        <input
          name="city"
          onChange={[Function]}
          placeholder="Your city"
          style={
            Object {
              "color": "#e74c3c",
            }
          }
          type="text"
          value=""
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#e74c3c",
            }
          }
        >
          Postal Code
        </label>
        <input
          name="postalCode"
          onChange={[Function]}
          placeholder="00-000"
          style={
            Object {
              "color": "#e74c3c",
            }
          }
          type="text"
          value=""
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#e74c3c",
            }
          }
        >
          Street
        </label>
        <input
          name="street"
          onChange={[Function]}
          placeholder="Street"
          style={
            Object {
              "color": "#e74c3c",
            }
          }
          type="text"
          value=""
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#e74c3c",
            }
          }
        >
          House / apartment number 
        </label>
        <input
          name="houseNumber"
          onChange={[Function]}
          placeholder="Number"
          style={
            Object {
              "color": "#e74c3c",
            }
          }
          type="text"
          value=""
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
      <div
        className="sc-jrcTuL cwMVBS"
      >
        <label
          style={
            Object {
              "color": "#e74c3c",
            }
          }
        >
          Phone
        </label>
        <input
          name="phone"
          onChange={[Function]}
          placeholder="000-000-000"
          style={
            Object {
              "color": "#e74c3c",
            }
          }
          type="tel"
          value=""
        />
        <p
          className="paragraph"
          style={
            Object {
              "color": "#e74c3c",
              "fontFamily": "Source Sans Pro",
              "fontSize": "1rem",
              "fontWeight": "400",
              "letterSpacing": "0",
              "lineHeight": "1.6",
            }
          }
        />
      </div>
    </div>
  </form>
  <div
    className="sc-dmctIk eVtxMl"
  >
    <button
      className="sc-bcXHqe sc-gKPRtg sc-iBYQkv feNVH hUzBlu YZIpJ button-text-light btn-lg button"
      onClick={[Function]}
      style={
        Object {
          "fontSize": "1.05rem",
          "padding": "0.65rem 1.25rem",
        }
      }
    >
      Back
    </button>
    <button
      className="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-lg button"
      onClick={[Function]}
      style={
        Object {
          "fontSize": "1.05rem",
          "padding": "0.65rem 1.25rem",
        }
      }
    >
      Summary
    </button>
  </div>
</section>
`;
